In [342]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
import MDAnalysis as mda 
import MDAnalysisTests as mdt 
import nglview as nv 
from MDAnalysis.analysis import rms 
from MDAnalysis.analysis import align 
from MDAnalysis.analysis import distances 
import pdbtools as pdb 

## 3HB3

In [260]:
_3hb3 = mda.Universe('x-ray/O-state/3hb3.pdb')
_3hb3.atoms
#Check for atoms where the masses could not be guessed. (Not guessed masses are 0)
#np.where(_3hb3.atoms.masses==0)[0]
#check how to access MN atom
mn_3hb3 = _3hb3.select_atoms('resname MN')
#how do i change the mass of the Mn atom?

/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: MN
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))


In [261]:
heme_fe = _3hb3.select_atoms('name FE')
print(heme_fe)

<AtomGroup [<Atom 7957: FE of type FE of resname HEA, resid 559 and segid A and altLoc >, <Atom 8017: FE of type FE of resname HEA, resid 560 and segid A and altLoc >]>


In [262]:
_3hb3_heme_a_fe = _3hb3.select_atoms('name FE and resid 559')
_3hb3_heme_a3_fe = _3hb3.select_atoms('name FE and resid 560')
_3hb3_mn = _3hb3.select_atoms('name MN')
print(_3hb3_heme_a_fe, _3hb3_heme_a3_fe, _3hb3_mn)

<AtomGroup [<Atom 7957: FE of type FE of resname HEA, resid 559 and segid A and altLoc >]> <AtomGroup [<Atom 8017: FE of type FE of resname HEA, resid 560 and segid A and altLoc >]> <AtomGroup [<Atom 8078: MN of type MN of resname MN, resid 562 and segid A and altLoc >]>


## 1AR1

In [263]:
_1ar1 = mda.Universe('x-ray/O-state/1ar1.pdb')

In [264]:
heme_fe = _1ar1.select_atoms('name FE')
print(heme_fe)

<AtomGroup [<Atom 7926: FE of type FE of resname HEA, resid 562 and segid A and altLoc >, <Atom 7986: FE of type FE of resname HEA, resid 563 and segid A and altLoc >]>


In [265]:
_1ar1_heme_a_fe = _1ar1.select_atoms('name FE and resid 562')
_1ar1_heme_a3_fe = _1ar1.select_atoms('name FE and resid 563')
_1ar1_mn = _1ar1.select_atoms('name MG')
print(_1ar1_heme_a_fe, _1ar1_heme_a3_fe, _1ar1_mn)

<AtomGroup [<Atom 7926: FE of type FE of resname HEA, resid 562 and segid A and altLoc >]> <AtomGroup [<Atom 7986: FE of type FE of resname HEA, resid 563 and segid A and altLoc >]> <AtomGroup [<Atom 7924: MG of type MG of resname MG, resid 560 and segid A and altLoc >]>


## 7AU3

In [266]:
_7au3 = mda.Universe('cEM/F-state/7au3.pdb')
#np.where(_7au3.atoms.masses==0)[0]
mn_7au3 = _7au3.select_atoms('resname MN')

/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/MDAnalysis/coordinates/PDB.py:429: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.
  warnings.warn("1 A^3 CRYST1 record,"


unit cell dimensions placeholder Warning can safely be ignored, 
since unit cell dimensions specified in the PDB are (1.0,1.0,1.0,90,90,90)

In [267]:
heme_fe = _7au3.select_atoms('name FE')
print(heme_fe)

<AtomGroup [<Atom 8673: FE of type FE of resname HEA, resid 602 and segid A and altLoc >, <Atom 8733: FE of type FE of resname HEA, resid 603 and segid A and altLoc >]>


In [268]:
_7au3_heme_a_fe = _7au3.select_atoms('name FE and resid 602')
_7au3_heme_a3_fe = _7au3.select_atoms('name FE and resid 603')
_7au3_mn = _7au3.select_atoms('name MN')
print(_7au3_heme_a_fe, _7au3_heme_a3_fe, _7au3_mn)

<AtomGroup [<Atom 8673: FE of type FE of resname HEA, resid 602 and segid A and altLoc >]> <AtomGroup [<Atom 8733: FE of type FE of resname HEA, resid 603 and segid A and altLoc >]> <AtomGroup [<Atom 8672: MN of type MN of resname MN, resid 601 and segid A and altLoc >]>


## 7AU6

In [269]:
_7au6 = mda.Universe('cEM/O-state/7au6.pdb')
#np.where(_7au6.atoms.masses==0)[0]
#mn_7au6 = _7au6.select_atoms('resname MN')

In [270]:
heme_fe = _7au6.select_atoms('name FE')
print(heme_fe)

<AtomGroup [<Atom 8643: FE of type FE of resname HEA, resid 601 and segid A and altLoc >, <Atom 8703: FE of type FE of resname HEA, resid 602 and segid A and altLoc >]>


In [271]:
_7au6_heme_a_fe = _7au6.select_atoms('name FE and resid 601')
_7au6_heme_a3_fe = _7au6.select_atoms('name FE and resid 602')
_7au6_mn = _7au6.select_atoms('name MN')
print(_7au6_heme_a_fe, _7au6_heme_a3_fe, _7au6_mn)

<AtomGroup [<Atom 8643: FE of type FE of resname HEA, resid 601 and segid A and altLoc >]> <AtomGroup [<Atom 8703: FE of type FE of resname HEA, resid 602 and segid A and altLoc >]> <AtomGroup [<Atom 8779: MN of type MN of resname MN, resid 612 and segid A and altLoc >]>


## 7ATE

In [272]:
_7ate = mda.Universe('cEM/P-state/7ate.pdb')
#np.where(_7ate.atoms.masses==0)[0]
#mn_7ate = _7ate.select_atoms('resname MN')

In [273]:
heme_fe = _7ate.select_atoms('name FE')
print(heme_fe)

<AtomGroup [<Atom 8693: FE of type FE of resname HEA, resid 602 and segid A and altLoc >, <Atom 8753: FE of type FE of resname HEA, resid 603 and segid A and altLoc >]>


In [274]:
_7ate_heme_a_fe = _7ate.select_atoms('name FE and resid 602')
_7ate_heme_a3_fe = _7ate.select_atoms('name FE and resid 603')
_7ate_mn = _7ate.select_atoms('name MN')
print(_7ate_heme_a_fe, _7ate_heme_a3_fe, _7ate_mn)

<AtomGroup [<Atom 8693: FE of type FE of resname HEA, resid 602 and segid A and altLoc >]> <AtomGroup [<Atom 8753: FE of type FE of resname HEA, resid 603 and segid A and altLoc >]> <AtomGroup [<Atom 8692: MN of type MN of resname MN, resid 601 and segid A and altLoc >]>


## 7ATN

In [275]:
_7atn = mda.Universe('cEM/R-state/7atn.pdb')
#np.where(_7atn.atoms.masses==0)[0]
#mn_7atn = _7atn.select_atoms('resname MN')

In [276]:
heme_fe = _7atn.select_atoms('name FE')
print(heme_fe)

<AtomGroup [<Atom 8691: FE of type FE of resname HEA, resid 602 and segid A and altLoc >, <Atom 8753: FE of type FE of resname HEA, resid 605 and segid A and altLoc >]>


In [277]:
_7atn_heme_a_fe = _7atn.select_atoms('name FE and resid 602')
_7atn_heme_a3_fe = _7atn.select_atoms('name FE and resid 603')
_7atn_mn = _7atn.select_atoms('name MN')
print(_7atn_heme_a_fe, _7atn_heme_a3_fe, _7atn_mn)

<AtomGroup [<Atom 8691: FE of type FE of resname HEA, resid 602 and segid A and altLoc >]> <AtomGroup []> <AtomGroup [<Atom 8690: MN of type MN of resname MN, resid 601 and segid A and altLoc >]>


## Distances

In [278]:
dist_3hb3 = distances.self_distance_array(_3hb3.select_atoms('name FE or name MN or name CU').positions)
dist_3hb3
#iron(heme a) - iron(heme a3), iron(heme a) - MN, iron(heme a3) - MN, 270 CU - iron heme a3, 271 Cu - iron heme a3, 
# Cu 561 - iron heme a, 560 fe - mn, 560 fe 270 cu, 560 fe 271 cu, 561 cu - mn
# Cu 561 - 270 Cu, 561 cu - 271 cu, mn - 270 cu, mn - 271 cu, 270 cu - 271 cu

array([13.27784298, 15.89315557, 17.05671864,  4.62071264, 14.11816285,
       13.19991472])

In [279]:
print(_3hb3.select_atoms('name FE or name MN'))

<AtomGroup [<Atom 7957: FE of type FE of resname HEA, resid 559 and segid A and altLoc >, <Atom 8017: FE of type FE of resname HEA, resid 560 and segid A and altLoc >, <Atom 8078: MN of type MN of resname MN, resid 562 and segid A and altLoc >]>


Find out which distance in self_distance_array is which, when selection worded like above

In [280]:
fe_3hb3 = _3hb3.select_atoms('name FE')
mn_3hb3 = _3hb3.select_atoms('name MN')
distances.distance_array(fe_3hb3.positions, mn_3hb3.positions, box=_3hb3.dimensions)
#distances.distance_array(fe_3hb3.positions, fe_3hb3.positions, box=_3hb3.dimensions)

array([[17.05671836],
       [14.11816283]])

In [281]:
np.append(np.array('O'), dist_3hb3)

array(['O', '13.277842982266828', '15.893155574384092',
       '17.056718639586112', '4.620712644300235', '14.11816284634431',
       '13.199914724918884'], dtype='<U32')

In [282]:
#iron(heme a) - iron(heme a3), iron(heme a) - MN, iron(heme a3) - MN 
dist_3hb3 = distances.self_distance_array(_3hb3.select_atoms('name FE or name MN or name CU').positions)
dist_3hb3 = np.append(np.array('O'), dist_3hb3)

dist_1ar1 = distances.self_distance_array(_1ar1.select_atoms('name FE or name MG or name CU').positions)
dist_1ar1 = np.append(np.array('O'), dist_1ar1)

dist_7au3 = distances.self_distance_array(_7au3.select_atoms('name FE or name MN or name CU or name CU1 or name CU2').positions)
dist_7au3 = np.append(np.array('F'), dist_7au3)

dist_7au6 = distances.self_distance_array(_7au6.select_atoms('name FE or name MN or name CU or name CU1 or name CU2').positions)
dist_7au6 = np.append(np.array('O'), dist_7au6)

dist_7ate = distances.self_distance_array(_7ate.select_atoms('name FE or name MN or name CU or name CU1 or name CU2').positions)
dist_7ate = np.append(np.array('P'), dist_7ate)

dist_7atn = distances.self_distance_array(_7atn.select_atoms('name FE or name MN or name CU or name CU1 or name CU2').positions)
dist_7atn = np.append(np.array('R'), dist_7atn)

#print(dist_3hb3, dist_1ar1, dist_7au3, dist_7au6, dist_7ate, dist_7atn)


In [283]:
dist = pd.DataFrame([dist_3hb3, dist_1ar1, dist_7au3, dist_7au6, dist_7ate, dist_7atn], index=['3HB3', '1AR1', '7AU3', '7AU6', '7ATE', '7ATN'], columns=['state', 'Fe(a) - Fe(a3)', 'Fe(a) - Mn/Mg', 'Fe(a3) - Mn/Mg', 'Cu_A1 - Fe(a3)', 'Cu_A2 - Fe(a3)', 'Cu_B - Fe(a)', 'Fe(a3) - MG/MN', 'Fe(a3) - Cu_A1', 'Fe(a3) - Cu_A2', 'Cu_B - Mn/Mg', 'Cu_B - Cu_A1', 'Cu_B - Cu_A2', 'Mn/Mg - Cu_A2', 'Mn/Mg - Cu_A1', 'Cu_A1 - Cu_A2'])
dist

state      Fe(a) - Fe(a3)       Fe(a) - Mn/Mg      Fe(a3) - Mn/Mg  \
3HB3     O  13.277842982266828  15.893155574384092  17.056718639586112   
1AR1     O  13.473607189592569  15.338440392860019   4.455524211906863   
7AU3     F  16.796761757479654  14.212012782785207  13.183533897294945   
7AU6     O  13.198579903539796   16.02582739845232  16.874577057804608   
7ATE     P   16.81747375569444  13.967247672867101  12.961699045571441   
7ATN     R   16.85996414081963  13.085282722169797  14.190758540856377   

          Cu_A1 - Fe(a3)      Cu_A2 - Fe(a3)        Cu_B - Fe(a)  \
3HB3   4.620712644300235   14.11816284634431  13.199914724918884   
1AR1  16.580954678405433   14.28779573433909    13.1413717762489   
7AU3   9.319262184914082  10.710414343356609  13.360880458079311   
7AU6  19.496896034907856  22.072102246965642   4.660330508508928   
7ATE   9.180417385461864  10.892372019580153  13.228668431994347   
7ATN   9.582696199729295  10.593261990370793   15.25316430029518   

          Fe(a3) - MG/MN      Fe(a3) - Cu_A1      Fe(a3) - Cu_A2  \
3HB3                None                None                None   
1AR1                None                None                None   
7AU3  15.821935298101156   19.37285177915332  21.952174291712055   
7AU6  14.003897586360367  21.764737550327954  24.052710350848294   
7ATE  15.971221834436818  19.432038372352896   22.04446434926845   
7ATN  12.990196209532463    19.6947454986504  22.297626591515673   

            Cu_B - Mn/Mg        Cu_B - Cu_A1        Cu_B - Cu_A2  \
3HB3                None                None                None   
1AR1                None                None                None   
7AU3   4.462446620285363  22.134563898575674   24.23814571967684   
7AU6  13.003037726860024  21.498045572432424  23.518810400496378   
7ATE   4.673110652466655   21.85731357305884  24.161357898077792   
7ATN  3.8824061240160765  21.985083171306385   23.46443283111887   

          Mn/Mg - Cu_A2       Mn/Mg - Cu_A1       Cu_A1 - Cu_A2  
3HB3               None                None                None  
1AR1               None                None                None  
7AU3  21.91568959331642  23.669252353081784  2.7307145850803436  
7AU6  8.994717097129625  10.730920645393642   2.618770440494295  
7ATE  21.60939850980943  23.607814447024115   2.695562923425543  
7ATN  22.37758862102904  24.177510646493428  2.7494056943855307

In [284]:
#drop unrelevant distances
dist = dist.drop(['Cu_A1 - Fe(a3)', 'Cu_A2 - Fe(a3)', 'Fe(a3) - Cu_A1', 'Fe(a3) - Cu_A2', 'Cu_B - Cu_A1', 'Cu_B - Cu_A2', 'Mn/Mg - Cu_A2', 'Mn/Mg - Cu_A1', 'Cu_A1 - Cu_A2'], axis=1)
dist

state      Fe(a) - Fe(a3)       Fe(a) - Mn/Mg      Fe(a3) - Mn/Mg  \
3HB3     O  13.277842982266828  15.893155574384092  17.056718639586112   
1AR1     O  13.473607189592569  15.338440392860019   4.455524211906863   
7AU3     F  16.796761757479654  14.212012782785207  13.183533897294945   
7AU6     O  13.198579903539796   16.02582739845232  16.874577057804608   
7ATE     P   16.81747375569444  13.967247672867101  12.961699045571441   
7ATN     R   16.85996414081963  13.085282722169797  14.190758540856377   

            Cu_B - Fe(a)      Fe(a3) - MG/MN        Cu_B - Mn/Mg  
3HB3  13.199914724918884                None                None  
1AR1    13.1413717762489                None                None  
7AU3  13.360880458079311  15.821935298101156   4.462446620285363  
7AU6   4.660330508508928  14.003897586360367  13.003037726860024  
7ATE  13.228668431994347  15.971221834436818   4.673110652466655  
7ATN   15.25316430029518  12.990196209532463  3.8824061240160765

observe values not matching expectations: 
Fe(a3) - Mn/Mg: 1AR1
Cu_B - Fe(a): 1AR1,7AU3,7ATE
Cu_B - Mn/Mg: 7AU3, 7ATE

Order determined for 3HB3
iron(heme a) - iron(heme a3), iron(heme a) - MN, iron(heme a3) - MN, 270 CU - iron heme a3, 271 Cu - iron heme a3, 
Cu 561 - iron heme a, 560 fe - mn, 560 fe 270 cu, 560 fe 270 cu, 561 cu - mn
Cu 561 - 270 Cu, 561 cu - 271 cu, mn - 270 cu, mn - 271 cu, 270 cu - 271 cu

In [285]:
dist_7au3 = distances.self_distance_array(_7au3.select_atoms('name FE or name MN or name CU or name CU1 or name CU2').positions)
dist_7au3 = np.append(np.array('F'), dist_7au3)
dist_7au3

array(['F', '16.796761757479654', '14.212012782785207',
       '13.183533897294945', '9.319262184914082', '10.710414343356609',
       '13.360880458079311', '15.821935298101156', '19.37285177915332',
       '21.952174291712055', '4.462446620285363', '22.134563898575674',
       '24.23814571967684', '21.91568959331642', '23.669252353081784',
       '2.7307145850803436'], dtype='<U32')

Rename Copper Atoms in the PDB files, create copies of the original PDBs with the _original.pdb file-ending. Check rename-ing.

In [286]:
_7au3_ = mda.Universe('cEM/F-state/7au3.pdb')
CuA_7au3 = _7au3_.select_atoms('resname CUA')
CuB_7au3 = _7au3_.select_atoms('resname CUB')

/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/MDAnalysis/topology/guessers.py:146: UserWarning: Failed to guess the mass for the following atom types: MN
  warnings.warn("Failed to guess the mass for the following atom types: {}".format(atom_type))
/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/MDAnalysis/coordinates/PDB.py:429: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.
  warnings.warn("1 A^3 CRYST1 record,"


In [287]:
print(CuA_7au3)

<AtomGroup [<Atom 8798: CU1 of type CU of resname CUA, resid 301 and segid B and altLoc >, <Atom 8799: CU2 of type CU of resname CUA, resid 301 and segid B and altLoc >]>


In [288]:
print(CuB_7au3)

<AtomGroup [<Atom 8793: CU of type CU of resname CUB, resid 604 and segid A and altLoc >]>


In [289]:
_7au6_ = mda.Universe('cEM/O-state/7au6.pdb')
CuA_7au6 = _7au6_.select_atoms('resname CUA')
CuB_7au6 = _7au6_.select_atoms('resname CUB')

In [290]:
print(CuA_7au6)

<AtomGroup [<Atom 8780: CU1 of type CU of resname CUA, resid 301 and segid B and altLoc >, <Atom 8781: CU2 of type CU of resname CUA, resid 301 and segid B and altLoc >]>


In [291]:
print(CuB_7au6)

<AtomGroup [<Atom 8763: CU of type CU of resname CUB, resid 603 and segid A and altLoc >]>


In [292]:
_7ate_ = mda.Universe('cEM/P-state/7ate.pdb')
CuA_7ate = _7ate_.select_atoms('resname CUA')
CuB_7ate = _7ate_.select_atoms('resname CUB')

In [293]:
print(CuA_7ate)

<AtomGroup [<Atom 8817: CU1 of type CU of resname CUA, resid 301 and segid B and altLoc >, <Atom 8818: CU2 of type CU of resname CUA, resid 301 and segid B and altLoc >]>


In [294]:
print(CuB_7ate)

<AtomGroup [<Atom 8813: CU of type CU of resname CUB, resid 604 and segid A and altLoc >]>


In [295]:
_7atn_ = mda.Universe('cEM/R-state/7atn.pdb')
CuA_7atn = _7atn_.select_atoms('resname CUA')
CuB_7atn = _7atn_.select_atoms('resname CUB')

In [296]:
print(CuA_7atn)

<AtomGroup [<Atom 8813: CU1 of type CU of resname CUA, resid 301 and segid B and altLoc >, <Atom 8814: CU2 of type CU of resname CUA, resid 301 and segid B and altLoc >]>


In [297]:
print(CuB_7atn)

<AtomGroup [<Atom 8751: CU of type CU of resname CUB, resid 603 and segid A and altLoc >]>


In [298]:
_1ar1_ = mda.Universe('x-ray/O-state/1ar1.pdb')
CuA_1ar1 = _1ar1_.select_atoms('resname CUA')
CuB_1ar1 = _1ar1_.select_atoms('resname CUB')

In [299]:
print(CuA_1ar1)

<AtomGroup [<Atom 8142: CU1 of type CU of resname CUA, resid 270 and segid B and altLoc >, <Atom 8143: CU2 of type CU of resname CUA, resid 271 and segid B and altLoc >]>


In [300]:
print(CuB_1ar1)

<AtomGroup [<Atom 7923: CU of type CU of resname CUB, resid 559 and segid A and altLoc >]>


In [301]:
_3hb3_ = mda.Universe('x-ray/O-state/3hb3.pdb')
CuA_3hb3 = _3hb3_.select_atoms('resname CUA')
CuB_3hb3 = _3hb3_.select_atoms('resname CUB')

In [302]:
print(CuA_3hb3)

<AtomGroup [<Atom 8497: CU1 of type CU of resname CUA, resid 270 and segid B and altLoc >, <Atom 8498: CU2 of type CU of resname CUA, resid 271 and segid B and altLoc >]>


In [303]:
print(CuB_3hb3)

<AtomGroup [<Atom 8077: CU of type CU of resname CUB, resid 561 and segid A and altLoc >]>


According to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2638054/ the Mn is measured because Mg is not measurable in EPR measurements, and therefore the bacterria is grown on Manganese rich soil because that is measurable with EPR techniques and not affecting function that much.

"Rhodobacter sphaeroides grown in a Mn(II)-rich medium replaces the intrinsic Mg(II) ion with an EPR-detectable Mn(II) ion without change in activity. Due to its close proximity and a shared ligand, oxidized CuA is spin-coupled to the Mn(II) ion, affecting the EPR spectrum. An examination of both bovine and R.s. oxidase crystal structures reveals a hydrogen-bonding pattern in the vicinity of the Mg(II) site that is consistent with three water ligands of the Mg(Mn) center when CuA is oxidized."

Therefore, all instances of MN will be changed to MG in the modified PDB files. Reload Files with Mg atoms. 

In [307]:
_3hb3 = mda.Universe('x-ray/O-state/3hb3.pdb')
_1ar1 = mda.Universe('x-ray/O-state/1ar1.pdb')
_7au3 = mda.Universe('cEM/F-state/7au3.pdb')
_7au6 = mda.Universe('cEM/O-state/7au6.pdb')
_7ate = mda.Universe('cEM/P-state/7ate.pdb')
_7atn = mda.Universe('cEM/R-state/7atn.pdb')

Add Dimensions to 7(...) structures, since none are given with the PDB files

In [308]:
max(_7au3.atoms.positions[:, 0]) - min(_7au3.atoms.positions[:, 0])

75.113

In [314]:
_7au3.dimensions=[93.500,151.000,156.700,90.00,90.00,90.00]
_7au3.dimensions
ngl.show_mdanalysis(_7au3)

/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/MDAnalysis/coordinates/PDB.py:429: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.
  warnings.warn("1 A^3 CRYST1 record,"


NGLWidget()

Warning:

/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/MDAnalysis/coordinates/PDB.py:429: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.
  warnings.warn("1 A^3 CRYST1 record,"

Changing dimensions with MDA does not work properly, so I changed the unit cell dimensions (CRYST1 in the PDB files) to the CRYST1 values of 1AR1, which are very similar to the CRYST1 values of 3HB3. 

In [315]:
_3hb3 = mda.Universe('x-ray/O-state/3hb3.pdb')
_1ar1 = mda.Universe('x-ray/O-state/1ar1.pdb')
_7au3 = mda.Universe('cEM/F-state/7au3.pdb')
_7au6 = mda.Universe('cEM/O-state/7au6.pdb')
_7ate = mda.Universe('cEM/P-state/7ate.pdb')
_7atn = mda.Universe('cEM/R-state/7atn.pdb')

## Alignment

Backbone and C-alpha alignment to check for easily overlooked mistakes

In [319]:
#merged_xray = mda.Merge(_3hb3.atoms, _1ar1.atoms)
#nv.show_mdanalysis(merged_xray)
_3hb3_1ar1 = align.alignto(_3hb3, _1ar1, select='backbone or name CA', match_atoms=True)
print(_3hb3_1ar1)
#nv.show_mdanalysis(mda.Merge(_3hb3.atoms, _1ar1.atoms))
RMSD = rms.rmsd(_3hb3.select_atoms('backbone').positions, _1ar1.select_atoms('backbone').positions, center=True, superposition=True)
print(f'RMSD in Angstrom: {RMSD} between structures 3HB3 and 1AR1')

(0.5560523516039728, 0.556052351601856)
RMSD in Angstrom: 0.5558103509757741 between structures 3HB3 and 1AR1


In [321]:
#_3hb3_7au3 = align.alignto(_3hb3, _7au3, select='backbone or name CA', match_atoms=True)
#print(_3hb3_7au3)
#RMSD = rms.rmsd(_3hb3.select_atoms('backbone').positions, _7au3.select_atoms('backbone').positions, center=True, superposition=True)
#print(f'RMSD in Angstrom: {RMSD} between structures 3HB3 and 7AU3')

Distance between backbone atoms present between x-ray structures and cEM structure need different selection when aligning and comparing

In [339]:
_3hb3_resids = _3hb3.select_atoms('backbone or name CA').residues
_7au3_resids = _7au3.select_atoms('backbone or name CA').residues
diff = list(set(list(_3hb3_resids)).difference(list(_7au3_resids)))
diff

[]

In [337]:
_3hb3_resids = _3hb3.select_atoms('backbone or name CA').residues
list(_3hb3_resids)

[<Residue GLY, 17>,
 <Residue PHE, 18>,
 <Residue PHE, 19>,
 <Residue THR, 20>,
 <Residue ARG, 21>,
 <Residue TRP, 22>,
 <Residue PHE, 23>,
 <Residue MET, 24>,
 <Residue SER, 25>,
 <Residue THR, 26>,
 <Residue ASN, 27>,
 <Residue HIS, 28>,
 <Residue LYS, 29>,
 <Residue ASP, 30>,
 <Residue ILE, 31>,
 <Residue GLY, 32>,
 <Residue ILE, 33>,
 <Residue LEU, 34>,
 <Residue TYR, 35>,
 <Residue LEU, 36>,
 <Residue PHE, 37>,
 <Residue THR, 38>,
 <Residue ALA, 39>,
 <Residue GLY, 40>,
 <Residue ILE, 41>,
 <Residue VAL, 42>,
 <Residue GLY, 43>,
 <Residue LEU, 44>,
 <Residue ILE, 45>,
 <Residue SER, 46>,
 <Residue VAL, 47>,
 <Residue CYS, 48>,
 <Residue PHE, 49>,
 <Residue THR, 50>,
 <Residue VAL, 51>,
 <Residue TYR, 52>,
 <Residue MET, 53>,
 <Residue ARG, 54>,
 <Residue MET, 55>,
 <Residue GLU, 56>,
 <Residue LEU, 57>,
 <Residue GLN, 58>,
 <Residue HIS, 59>,
 <Residue PRO, 60>,
 <Residue GLY, 61>,
 <Residue VAL, 62>,
 <Residue GLN, 63>,
 <Residue TYR, 64>,
 <Residue MET, 65>,
 <Residue CYS, 66>,


Comparing structures does not work because comparing lists of Residue objects does not work with regular pythonic ways. Therefore, the structures will be changed in a way, that chain names are more clear, and can be accessed more easily and less error prone. 

PDB Format:

1 -  6         Record name   "ATOM  "

7 - 11         Integer       serial       Atom  serial number.


13 - 16        Atom          name         Atom name.


17             Character     altLoc       Alternate location indicator.


18 - 20        Residue name  resName      Residue name.


22             Character     chainID      Chain identifier.


23 - 26        Integer       resSeq       Residue sequence number.


27             AChar         iCode        Code for insertion of residues.


31 - 38        Real(8.3)     x            Orthogonal coordinates for X in Angstroms.


39 - 46        Real(8.3)     y            Orthogonal coordinates for Y in Angstroms.


47 - 54        Real(8.3)     z            Orthogonal coordinates for Z in Angstroms.


55 - 60        Real(6.2)     occupancy    Occupancy.


61 - 66        Real(6.2)     tempFactor   Temperature  factor.


77 - 78        LString(2)    element      Element symbol, right-justified.


79 - 80        LString(2)    charge       Charge  on the atom.


Using library pdb-tools to edit PDB files. 

In [353]:
from pdbtools import pdb_wc
from pdbtools import pdb_delchain

Syntax for use inside python programs not given in the program overview, or the github overview, so here it is:

Syntax: info = function.checkinput(pdb-file)

Syntax: variable = function.run(pdb-file, option='m')

Example Syntax: info = pdb_wc.checkinput('x-ray/O-state/1ar1.pdb')

Example Syntax: info = pdb_wc.run('x-ray/O-state/1ar1.pdb', option='m')

In [363]:
#_7au3_new = pdb_delchain.check_input('cEM/F-state/7au3.pdb')#, ('C','D'))


Deletes all atoms matching specific chains in the PDB file.

Usage:
    python pdb_delchain.py -<option> <pdb file>

Example:
    python pdb_delchain.py -A 1CTF.pdb  # removes chain A from PDB file
    python pdb_delchain.py -A,B 1CTF.pdb  # removes chains A and B from PDB file

This program is part of the `pdb-tools` suite of utilities and should not be
distributed isolatedly. The `pdb-tools` were created to quickly manipulate PDB
files using the terminal, and can be used sequentially, with one tool streaming
data to another. They are based on old FORTRAN77 code that was taking too much
effort to maintain and compile. RIP.


SystemExit: 1

/home/pbuser/miniconda3/envs/work/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3452: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


In [364]:
#print(_7au3_new)

<generator object run at 0x7fc59ca81cf0>


Since many functions from pdb-tools do not work inside a jupyter notebook, we compile a bash script that does the analyis for us and then execute the bash file that calls on all the python files. The package has to be in a subdirectory of the folder where execution happens though. 